components/brave_shields/browser/domain_block_navigation_throttle.cc

```diff
@@ -0,0 +1,165 @@
+  std::string mock_data_url;
+  ad_block_service->ShouldStartRequest(
+      url, blink::mojom::ResourceType::kMainFrame, url.host(), &did_match_rule,
+      &did_match_exception, &did_match_important, &mock_data_url);
+  return (did_match_important || (did_match_rule && !did_match_exception));
+}
+
+}  // namespace
+
+namespace brave_shields {
+
+// static
+std::unique_ptr<DomainBlockNavigationThrottle>
+DomainBlockNavigationThrottle::MaybeCreateThrottleFor(
+    content::NavigationHandle* navigation_handle,
+    AdBlockService* ad_block_service,
+    AdBlockCustomFiltersService* ad_block_custom_filters_service,
+    const std::string& locale) {
+  if (!ad_block_service || !ad_block_custom_filters_service)
+    return nullptr;
+  return std::make_unique<DomainBlockNavigationThrottle>(
+      navigation_handle, ad_block_service, ad_block_custom_filters_service,
+      locale);
+}
+
+DomainBlockNavigationThrottle::DomainBlockNavigationThrottle(
+    content::NavigationHandle* navigation_handle,
+    AdBlockService* ad_block_service,
+    AdBlockCustomFiltersService* ad_block_custom_filters_service,
+    const std::string& locale)
+    : content::NavigationThrottle(navigation_handle),
+      ad_block_service_(ad_block_service),
+      ad_block_custom_filters_service_(ad_block_custom_filters_service),
+      locale_(locale) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+}
+
+DomainBlockNavigationThrottle::~DomainBlockNavigationThrottle() {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+}
+
+content::NavigationThrottle::ThrottleCheckResult
+DomainBlockNavigationThrottle::WillStartRequest() {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+
+  // Don't bother checking non-HTTP(S) pages
+  content::NavigationHandle* handle = navigation_handle();
+  GURL request_url = handle->GetURL();
+  if (!request_url.SchemeIsHTTPOrHTTPS())
+    return content::NavigationThrottle::PROCEED;
+
+  // If user has just chosen to proceed on our interstitial, don't show
+  // another one.
+  content::WebContents* web_contents = handle->GetWebContents();
+  DomainBlockTabStorage* tab_storage =
+      DomainBlockTabStorage::GetOrCreate(web_contents);
+  if (tab_storage->IsProceeding())
+    return content::NavigationThrottle::PROCEED;
+
+  // Otherwise, call the ad block service on a task runner to determine whether
+  // this domain should be blocked.
+  ad_block_service_->GetTaskRunner()->PostTaskAndReplyWithResult(
+      FROM_HERE,
+      base::BindOnce(&ShouldBlockDomainOnTaskRunner, ad_block_service_,
+                     request_url),
+      base::BindOnce(&DomainBlockNavigationThrottle::OnShouldBlockDomain,
+                     weak_ptr_factory_.GetWeakPtr()));
+
+  // Since the call to the ad block service is asynchronous, we defer the final
+  // decision of whether to allow or block this navigation. The callback from
+  // the task runner will call a method to give our final answer.
+  return content::NavigationThrottle::DEFER;
```

Comment:
I'd suggest to follow an SB pattern: it allows the request to proceed until headers are received, and then defers if the check is still in progress

Reply:
The entire purpose of this is to deny any network requests to the first-party domain until the user agrees to proceed.